test/e2e/gardener/shoot/internal/inclusterclient/inclusterclient.go

```diff
@@ -0,0 +1,281 @@
+// SPDX-FileCopyrightText: SAP SE or an SAP affiliate company and Gardener contributors
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package inclusterclient
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"maps"
+	"time"
+
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+	. "github.com/onsi/gomega/gbytes"
+	"github.com/onsi/gomega/gexec"
+	corev1 "k8s.io/api/core/v1"
+	rbacv1 "k8s.io/api/rbac/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/gardener/gardener/imagevector"
+	gardencorev1beta1 "github.com/gardener/gardener/pkg/apis/core/v1beta1"
+	v1beta1constants "github.com/gardener/gardener/pkg/apis/core/v1beta1/constants"
+	resourcesv1alpha1 "github.com/gardener/gardener/pkg/apis/resources/v1alpha1"
+	"github.com/gardener/gardener/pkg/client/kubernetes"
+	imagevectorutils "github.com/gardener/gardener/pkg/utils/imagevector"
+	"github.com/gardener/gardener/test/framework"
+)
+
+const (
+	name      = "e2e-test-in-cluster-client"
+	namespace = metav1.NamespaceDefault
+
+	podNameDirect                 = name + "-direct"
+	podNameAPIServerProxyIP       = name + "-apiserver-proxy-ip"
+	podNameAPIServerProxyHostname = name + "-apiserver-proxy-hostname"
+	containerName                 = "kubectl"
+)
+
+var labels = map[string]string{"e2e-test": "in-cluster-client"}
+
+// VerifyInClusterAccessToAPIServer verifies access to the shoot API server from in-cluster clients.
+// It verifies the following paths:
+// - direct path using the KUBERNETES_SERVICE_HOST env var injected by gardener
+// - via the API server proxy's clusterIP
+// - via the API server proxy's hostname kubernetes.default.svc
```

Comment:
Nit: Technically, it seems like the full name is used instead.

```suggestion
// - via the API server proxy's hostname kubernetes.default.svc.cluster.local
```